airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/test/util/destination_process/DockerizedDestination.kt

```diff
@@ -106,20 +123,89 @@ class DockerizedDestination(
+                                null, // this should be impossible, treat it as error
+                                AirbyteLogMessage.Level.FATAL, // klogger doesn't have a fatal level
+                                AirbyteLogMessage.Level.ERROR -> logger.error { combinedMessage }
+                                AirbyteLogMessage.Level.WARN -> logger.warn { combinedMessage }
+                                AirbyteLogMessage.Level.INFO -> logger.info { combinedMessage }
+                                AirbyteLogMessage.Level.DEBUG -> logger.debug { combinedMessage }
+                                AirbyteLogMessage.Level.TRACE -> logger.trace { combinedMessage }
+                            }
+                        } else {
+                            destinationOutput.accept(message)
+                        }
+                    }
+                    stdoutDrained.complete(Unit)
+                }
+                launch {
+                    // Consume stderr. Connectors shouldn't really use this,
+                    // and whatever this stream contains, it's almost certainly not valid messages.
+                    // Dump it straight to our own stderr.
+                    process.errorReader().forEachLine { logger.error { it } }
+                    stderrDrained.complete(Unit)
+                }
+            }
+        }
     }
 
     override fun sendMessage(message: AirbyteMessage) {
-        // push a message to the docker process' stdin
-        TODO("Not yet implemented")
+        destinationStdin.write(Jsons.writeValueAsString(message))
+        destinationStdin.newLine()
     }
 
     override fun readMessages(): List<AirbyteMessage> {
-        // read everything from the process' stdout
-        TODO("Not yet implemented")
+        return destinationOutput.newMessages()
     }
 
     override fun shutdown() {
-        // close stdin, wait until process exits
-        TODO("Not yet implemented")
+        destinationStdin.close()
+        // The old cdk had a 1-minute timeout here. That seems... weird?
+        // We can just rely on the junit timeout, presumably?
+        process.waitFor()
+        // Wait for ourselves to drain stdout/stderr. Otherwise we won't capture
+        // all the destination output (logs/trace messages).
+        runBlocking {
+            stdoutDrained.join()
+            stderrDrained.join()
+        }
+        val exitCode = process.exitValue()
+        if (exitCode != 0) {
+            // Hey look, it's possible to extract the error from a failed destination process!
+            // because "destination exit code 1" is the least-helpful error message.
+            val filteredTraces =
+                destinationOutput.traces()
+                    .filter { it.type == AirbyteTraceMessage.Type.ERROR }
+                    .map { it.error }
+            throw RuntimeException(
+                """
+                    Destination process exited uncleanly: $exitCode
```

Comment:
FYI https://airbytehq-team.slack.com/archives/C03AS1GAQV6/p1728408552964509

Platform depends on exit non-zero, so the destination can never exit cleanly on failure.

Reply:
oh perfect. I think that works out in our favor - for 90% of tests, if the destination doesn't exit 0, we want the test to fail (and for the remaining tests, we can just wrap with assertThrows... though I should define a new exception class for this.)

Reply:
new exception class defined in https://github.com/airbytehq/airbyte/pull/46363/commits/7f443789d4685f7d32c88c22281438e59a9405de